geps/gep-1731/index.md

```diff
@@ -0,0 +1,409 @@
+    // from the gateway to a backend should be retried.
+    //
+    // If the maximum number of retries has been attempted without a successful
+    // response from the backend, the Gateway MUST return an error.
+    //
+    // When this field is unspecified, the number of times to attempt to retry
+    // a backend request is implementation-specific.
+    //
+    // Support: Extended
+    //
+    // +optional
+    Attempts *Int `json:"attempts,omitempty"`
+
+    // Backoff specifies the minimum duration a Gateway should wait between
+    // retry attempts and is represented in Gateway API Duration formatting.
+    //
+    // For example, setting the `rules[].retry.backoff` field to the value
+    // `100ms` will cause a backend request to first be retried approximately
+    // 100 milliseconds after timing out or reciveing a response code configured
+    // to be retryable.
+    //
+    // An implementation MAY use an exponential or alternative backoff strategy
+    // for subsequent retry attempts, MAY cap the maximum backoff duration to
+    // some amount greater than the specified minimum, and MAY add arbitrary
+    // jitter to stagger requests, as long as unsuccessful backend requests are
+    // not retried before the configured minimum duration.
+    //
+    // If a Request timeout (`rules[].timeouts.request`) is configured on the
+    // route, the entire duration of the initial request and any retry attempts
+    // MUST not exceed the Request timeout duration. If any retry attempts are
+    // still in progress when the Request timeout duration has been reached,
+    // these SHOULD be cancelled if possible and the Gateway MUST immediately
+    // return a timeout error.
+    //
+    // If a BackendRequest timeout (`rules[].timeouts.backendRequest`) is
+    // configured on the route, any retry attempts which reach the configured
+    // BackendRequest timeout duration without a response SHOULD be cancelled if
+    // possible and the Gateway should wait for at least the specified backoff
+    // duration before attempting to retry the backend request again.
+    //
+    // If a BackendRequest timeout is _not_ configured on the route, retry
+    // attempts MAY time out after an implementation default duration, or MAY
+    // remain pending until a configured Request timeout or implementation
+    // default duration for total request time is reached.
+    //
+    // When this field is unspecified, the time to wait between retry attempts
+    // is implementation-specific.
+    //
+    // Support: Extended
+    //
+    // +optional
+    Backoff *Duration `json:"backoff,omitempty"`
+}
+
+// HTTPRouteRetryStatusCode defines an HTTP response status code for
+// which a backend request should be retried.
+//
+// Implementations MUST support the following status codes as retryable:
+//
+// * "500"
+// * "502"
+// * "503"
+// * "504"
+//
+// Implementations SHOULD support the `"5xx"` shorthand for matching any
```

Comment:
It feels like the SHOULD and MUST are reversed here. IMO, it's likely going to be more portable to retry on any 5xx request than supporting specifying each 5xx code independently.

Reply:
It may be worth defining a couple Gateway values that we expect to be portable and have extended support:

- `5xx` for any 500-599 (inclusive) status
- `4xx` for any 400-499 (inclusive) status

We could potentially leave some wiggle room here where 5xx means at least 500, 502, 503, and 504, while 4xx means at least 400, 401, 403, 404, and 429. (Not completely sure on the list we'd want there, but specifying a minimum set of error codes that MUST be covered could help with portability).

Reply:
Retrying 400, 401, 403, 404, and 429 will almost always be a bad idea in practice -- there's a reason why Envoy only includes 409 in the `retriable-4xx` group. 🙂 I would honestly be OK with MUST support 5xx, MAY support 600-999, SHOULD NOT (but MAY) support 100-499. (IIRC, in Emissary-land we ran into one or two cases where someone really, really wanted to retry something that wasn't a 5yz, but only one or two cases. It was NOT common.)

Reply:
Haha yes, I think you're completely right. Apparently I was commenting a bit too late into my day/week. Agree with @kflynn's latest comment.

Reply:
Is retrying 429s really a bad idea? That code in particular seems like it should be retried, just with some additional logic/delay that I didn't want to get into here yet (and could possibly be part of a GEP for https://github.com/kubernetes-sigs/gateway-api/issues/326 at some point in the future).

Reply:
If you don't parse and honor `Retry-After`, or if there is no `Retry-After`, then retrying 429 is a disaster. 🙂  If you _do_ parse and honor `Retry-After`...

_&lt;semireligious&gt;_

...it's not a disaster but it's still a bad idea. 🙂 

1. `Retry-After` frequently specifies a largish interval (it's not uncommon to talk about rate limits in terms of "requests per hour" or "requests per day"). Chances are really good that the `Retry-After` interval will be much longer than the request timeout anyway.

2. The 4yz codes are really meant to say "hey client, you made a mistake" rather than "hey client, something went wrong on the server". In general, the client needs to do something different to clear the condition, so retrying without consulting the client is very rarely a good idea.

_&lt;/semireligious&gt;_

So... maybe let's not? It feels like a better idea to direct that effort elsewhere, to me anyway.

Reply:
> It feels like the SHOULD and MUST are reversed here. IMO, it's likely going to be more portable to retry on any 5xx request than supporting specifying each 5xx code independently.

While I agree with the relative importance of each of these from a Gateway API end user perspective (`"5xx"` is simpler and more useful), the SHOULD and MUST here was informed by existing proxy implementations of retry logic. I tightened up the scope of what 5xx means to hopefully allow more portability, and I'm open to moving it from SHOULD to MUST with support from additional non-Envoy implementations.

Reply:
> If you don't parse and honor Retry-After, or if there is no Retry-After, then retrying 429 is a disaster. 🙂 If you do parse and honor Retry-After...

Thinking back to when I've interacted with 429 codes previously, it was mostly rate limited APIs informing a script or client that a limit was exceeded, and the resolution required some manual intervention like rotating API keys, rotating IP addresses or subscribing to a higher-tier plan so yea this makes more sense to me now that implementing `Retry-After` support wouldn't be the highest priority for a gateway relaying requests.